janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java

```diff
@@ -780,19 +780,25 @@ private String getAdditionScript(KeyInformation.IndexRetriever information, Stri
         final StringBuilder script = new StringBuilder();
         for (final IndexEntry e : mutation.getAdditions()) {
             final KeyInformation keyInformation = information.get(storeName).get(e.field);
-            switch (keyInformation.getCardinality()) {
-                case SET:
-                case LIST:
-                    script.append("if(ctx._source[\"").append(e.field).append("\"] == null) ctx._source[\"").append(e.field).append("\"] = [];");
-                    script.append("ctx._source[\"").append(e.field).append("\"].add(").append(convertToJsType(e.value, compat.scriptLang(), Mapping.getMapping(keyInformation))).append(");");
-                    if (hasDualStringMapping(keyInformation)) {
-                        script.append("if(ctx._source[\"").append(getDualMappingName(e.field)).append("\"] == null) ctx._source[\"").append(getDualMappingName(e.field)).append("\"] = [];");
-                        script.append("ctx._source[\"").append(getDualMappingName(e.field)).append("\"].add(").append(convertToJsType(e.value, compat.scriptLang(), Mapping.getMapping(keyInformation))).append(");");
-                    }
-                    break;
-                default:
-                    break;
+            final Cardinality cardinality = keyInformation.getCardinality();
+
+            if (cardinality != Cardinality.SET && cardinality != Cardinality.LIST)
+                continue;
+
+            script.append("if(ctx._source[\"").append(e.field).append("\"] == null) ctx._source[\"").append(e.field).append("\"] = [];");
+            script.append("ctx._source[\"").append(e.field).append("\"].add(").append(convertToJsType(e.value, compat.scriptLang(), Mapping.getMapping(keyInformation))).append(");");
+
+            if (cardinality == Cardinality.SET)
+                script.append("ctx._source[\"").append(e.field).append("\"] = ").append("ctx._source[\"").append(e.field).append("\"].stream().distinct().collect(Collectors.toList());");
+
+            if (hasDualStringMapping(keyInformation)) {
+                final String dualFieldName = getDualMappingName(e.field);
+
+                script.append("if(ctx._source[\"").append(dualFieldName).append("\"] == null) ctx._source[\"").append(dualFieldName).append("\"] = [];");
+                script.append("ctx._source[\"").append(dualFieldName).append("\"].add(").append(convertToJsType(e.value, compat.scriptLang(), Mapping.getMapping(keyInformation))).append(");");
 
+                if (cardinality == Cardinality.SET)
+                    script.append("ctx._source[\"").append(dualFieldName).append("\"] = ").append("ctx._source[\"").append(dualFieldName).append("\"].stream().distinct().collect(Collectors.toList());");
```

Comment:
This like looks like a duplicate like of 792 line. Maybe it is better to create a separate method? Something like:
```
private void appendDistinct(StringBuilder script, String fieldName){
    script.append("ctx._source[\"").append(fieldName).append("\"] = ").append("ctx._source[\"").append(fieldName).append("\"].stream().distinct().collect(Collectors.toList());");
}
```

Reply:
Done. Initially I've thought about creating a new function but decided not to add it just for performance reasons. But actually I do not think it does really matter. 

If your point here is code readability then I'd like to suggest to rewrite it as follows
```java
        script.append(
            String.format(
                "ctx._source[\"%1$s\"] = ctx._source[\"%1$s\"].stream().distinct().collect(Collectors.toList());",
                fieldName
            )
        );
```

What do you think about it?

Reply:
> Initially I've thought about creating a new function but decided not to add it just for performance reasons

What do you mean by that? I thought Java compilers and JVM optimize method calls. Am I wrong?

> If your point here is code readability then I'd like to suggest to rewrite it as follows

I don't think we should sacrifice performance for code readability. I wouldn't use regular expression here because it is often too slow.

Reply:
> What do you mean by that? I thought Java compilers and JVM optimize method calls. Am I wrong?

You're right, JIT does method inlining. My bad.

> I don't think we should sacrifice performance for code readability. I wouldn't use regular expression here because it is often too slow.

I agree. But to tell the truth  I'm not sure if we really sacrifice here, since we're building a string with script that will be executed by ES. I assume executing script in ES is much slower then calling String.format. But ok, let's keep it as is.